File: Mis Prácticas/Napakalaki/src/napakalaki/NumericBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author fran
 */
public class NumericBadConsequence extends BadConsequence{
    
    // Variables
    private int nVisibleTreasures, nHiddenTreasures;
    
    // Constructor
    public NumericBadConsequence(String text, int levels, int nVisibleTreasures, 
            int nHiddenTreasures){
        
        super(text, levels);
        setNVisibleTreasures(nVisibleTreasures);
        setNHiddenTreasures(nHiddenTreasures);
        
    }
    
    // Getters
    public int getNVisibleTreasures(){
        return nVisibleTreasures;
    }
    
    public int getNHiddenTreasures(){
        return nHiddenTreasures;
    }
    
    // Setters
    public void setNVisibleTreasures(int nVisibleTreasures){
        this.nVisibleTreasures = nVisibleTreasures;
    }
    
    public void setNHiddenTreasures(int nHiddenTreasures){
        this.nHiddenTreasures = nHiddenTreasures;
    }
    
    // Utils
    @Override
    public boolean isEmpty(){
        
        return nVisibleTreasures==0 && nHiddenTreasures==0;
        
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        
        if(nVisibleTreasures>0)
            nVisibleTreasures--;
        
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        
        if(nHiddenTreasures>0)
            nHiddenTreasures--;
        
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(
        ArrayList<Treasure> visibleTreasures, 
        ArrayList<Treasure> hiddenTreasures){
        
        NumericBadConsequence bcAux = new NumericBadConsequence(super.getText(), 0, 0, 0);
        
        // Si el jugador tiene algún tesoro visible y la lista que recibe no
        // está vacía
        if(!visibleTreasures.isEmpty() && nVisibleTreasures>0){
                            
            // Si tiene nVisibleTreasures para descontar
            if(nVisibleTreasures>=visibleTreasures.size())
                bcAux.setNVisibleTreasures(visibleTreasures.size());
            else
                bcAux.setNVisibleTreasures(nVisibleTreasures);
     
        }
        
        // Si los tesoros ocultos que se reciben no están vacíos y si el jugador 
        // no tiene tesoros ocultos equipados y tiene algún tesoro oculto
        if(!hiddenTreasures.isEmpty() && nHiddenTreasures>0){
            
            //  Si tiene nHiddenTreasures para descontar
            if(nHiddenTreasures>=hiddenTreasures.size())
                bcAux.setNHiddenTreasures(hiddenTreasures.size());
            else
                bcAux.setNHiddenTreasures(nHiddenTreasures);
                 
        }
        
        return bcAux;
        
    }
    
    @Override
    public String toString(){
        
        return super.toString() +
            ", NVisibleTreasures = " + Integer.toString(nVisibleTreasures) +
            ", NHiddenTreasures = " + Integer.toString(nHiddenTreasures);      
        
    }
    
}

File: NapakalakiRuby/lib/numeric_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "bad_consequence.rb"

class NumericBadConsequence < BadConsequence
  
  public
  
  def initialize(t, l, v, h)
    super(t, l)
    @nVisibleTreasures = v
    @nHiddenTreasures = h
  end
  
  public_class_method :new
  
  # Getters
  def get_n_visible_treasures
    @nVisibleTreasures
  end
  
  def get_n_hidden_treasures
    @nHiddenTreasures
  end
  
  # Setters
  def set_n_visible_treasures(n)
    @nVisibleTreasures = n
  end
  
  def set_n_hidden_treasures(h)
    @nHiddenTreasures = h
  end
  
  # Utils
  def is_empty
    @nVisibleTreasures==0 && @nHiddenTreasures==0
  end
  
  def substract_visible_treasure(t)
    if @nVisibleTreasures>0
      @nVisibleTreasures = @nVisibleTreasures - 1
    end
  end
  
  def substract_hidden_treasure(t)
    if @nHiddenTreasures>0
      @nHiddenTreasures = @nHiddenTreasures - 1
    end
  end
  
  def adjust_to_fit_treasure_lists(v, h)
    
    bc_aux = NumericBadConsequence.new("", 0, 0, 0)
    
    # Si los tesoros visibles que se reciben no están vacíos
    if v.any? && @nVisibleTreasures>0

        # Si tiene nVisibleTreasures para descontar
        if @nVisibleTreasures>=v.length
          bc_aux.set_n_visible_treasures(v.length)
        else
          bc_aux.set_n_visible_treasures(@nVisibleTreasures)
        end
                
    end
    
    # Si los tesoros ocultos que se reciben no están vacíos
    if h.any? && @nHiddenTreasures>0

        # Si tiene nHiddenTreasures para descontar
        if @nHiddenTreasures>=h.length
          bc_aux.set_n_hidden_treasures(h.length)
        else
          bc_aux.set_n_hidden_treasures(@nHiddenTreasures)
        end
                  
    end
    
    bc_aux
    
  end
  
  def to_s
    super + ", NVisibleTreasures: #{@nVisibleTreasures}," +
        " NHiddenTreasures: #{@nHiddenTreasures}"
  end
    
end
